models/user_setting.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"strings"
+
+	"xorm.io/builder"
+)
+
+// UserSetting is a key value store of user settings
+type UserSetting struct {
+	ID     int64  `xorm:"pk autoincr"`
+	UserID int64  `xorm:"index"`              // to load all of someone's settings
+	Key    string `xorm:"varchar(255) index"` // ensure key is always lowercase
+	Value  string `xorm:"text"`
+}
+
+// BeforeInsert will be invoked by XORM before inserting a record
+func (setting *UserSetting) BeforeInsert() {
+	setting.Key = strings.ToLower(setting.Key)
+}
+
+// BeforeUpdate will be invoked by XORM before updating a record
+func (setting *UserSetting) BeforeUpdate() {
+	setting.Key = strings.ToLower(setting.Key)
+}
+
+// GetUserSetting returns specific settings from user
+func GetUserSetting(uid int64, keys []string) ([]*UserSetting, error) {
+	settings := make([]*UserSetting, 0, 5)
+	if err := x.
+		Where("user_id=?", uid).
+		And(builder.In("key", keys)).
+		Asc("id").
+		Find(&settings); err != nil {
+		return nil, err
+	}
+	return settings, nil
+}
+
+// GetUserAllSettings returns all settings from user
+func GetUserAllSettings(uid int64) ([]*UserSetting, error) {
+	settings := make([]*UserSetting, 0, 5)
+	if err := x.
+		Where("user_id=?", uid).
+		Asc("id").
+		Find(&settings); err != nil {
+		return nil, err
+	}
+	return settings, nil
+}
+
+func addUserSetting(e Engine, setting *UserSetting) error {
+	used, err := settingExists(e, setting.UserID, setting.Key)
+	if err != nil {
+		return err
+	} else if used {
+		return ErrUserSettingExists{setting}
+	}
+	_, err = e.Insert(setting)
+	return err
+}
+
+func settingExists(e Engine, uid int64, key string) (bool, error) {
+	if len(key) == 0 {
+		return true, nil
+	}
+
+	return e.Table(&UserSetting{}).Exist(&UserSetting{UserID: uid, Key: strings.ToLower(key)})
+}
+
+// DeleteUserSetting deletes a specific setting for a user
+func DeleteUserSetting(setting *UserSetting) error {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	if _, err := sess.Delete(setting); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+// SetUserSetting updates a users' setting for a specific key
+func SetUserSetting(setting *UserSetting) error {
+	err := addUserSetting(x, setting)
+	if err != nil && IsErrUserSettingExists(err) {
+		return updateUserSettingValue(x, setting)
+	}
+	return err
+}
+
+func updateUserSettingValue(e Engine, setting *UserSetting) error {
+	used, err := settingExists(e, setting.UserID, setting.Key)
+	if err != nil {
+		return err
```

Comment:
Why not just overwrite on update?

Reply:
I'm not sure what you mean, that's what this is doing, it checks if the toe exists and if it does then it will update it.

Reply:
I may have misread that. What does `settingExists` do? Does it check for existance of the row in the database, or does it check for the existance in a predefined list of settings?

Reply:
`settingExists` looks for a row that has both key & userid matching (to see if the key exists for a given user). In theory I could set a value for key `techknowlogick_is_cool` and userid `1337` and then `settingExists(e, 1337, techkn..)` would return true.

Reply:
I don't think this will always work. Depending on the transactional model, two transactions could be adding the key at the same time. Neither will see that the other one is inserting the key and so you will end up with two entries.

We would need to use an UPSERT pattern as above.